File: python-ref/gvibu_ref/commands/comm.py
"""comm: compare two sorted files line by line."""
import sys


def run(args: list[str]) -> int:
    flag1 = True
    flag2 = True
    flag3 = True

    files: list[str] = []
    for arg in args:
        if arg == "-1":
            flag1 = False
        elif arg == "-2":
            flag2 = False
        elif arg == "-3":
            flag3 = False
        elif arg.startswith('-') and len(arg) > 1:
            print(f"comm: invalid option: {arg}", file=sys.stderr)
            return 1
        else:
            files.append(arg)

    if len(files) < 2:
        print("comm: missing operand", file=sys.stderr)
        return 1
    if len(files) > 2:
        print(f"comm: extra operand: {files[2]}", file=sys.stderr)
        return 1

    try:
        fh1 = sys.stdin if files[0] == "-" else open(files[0])
        fh2 = sys.stdin if files[1] == "-" else open(files[1])
    except OSError as e:
        print(f"comm: {e}", file=sys.stderr)
        return 1

    line1 = None
    line2 = None

    with fh1 if files[0] != "-" else _noop(fh1) as f1, \
         fh2 if files[1] != "-" else _noop(fh2) as f2:
        while True:
            if line1 is None:
                line1 = f1.readline()
            if line2 is None:
                line2 = f2.readline()

            if not line1 and not line2:
                break

            if not line1:
                if flag2:
                    print(f"\t{line2.rstrip()}")
                line2 = None
                for l in f2:
                    if flag2:
                        print(f"\t{l.rstrip()}")
                break

            if not line2:
                if flag1:
                    print(f"{line1.rstrip()}")
                line1 = None
                for l in f1:
                    if flag1:
                        print(f"{l.rstrip()}")
                break

            l1 = line1.rstrip('\n\r')
            l2 = line2.rstrip('\n\r')

            if l1 < l2:
                if flag1:
                    print(l1)
                line1 = None
                line2 = line2  # keep
            elif l1 > l2:
                if flag2:
                    print(f"\t{l2}")
                line2 = None
                line1 = line1  # keep
            else:
                if flag3:
                    print(f"\t\t{l1}")
                line1 = None
                line2 = None

    return 0


class _noop:
    """Context manager for stdin (no-op close)."""
    def __init__(self, fh):
        self._fh = fh
    def __enter__(self):
        return self._fh
    def __exit__(self, *args):
        pass

File: rust/src/commands/comm.rs
/// comm: compare two sorted files line by line.
/// Output: column 1 = lines unique to file1, column 2 = lines unique to file2,
///         column 3 = lines common to both.
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};
use crate::pwriteln;

pub fn run(w: &mut dyn Write, args: &[String]) -> i32 {
    let mut flag1 = true; // suppress col 1
    let mut flag2 = true; // suppress col 2
    let mut flag3 = true; // suppress col 3

    let mut files: Vec<&str> = Vec::new();
    for arg in args {
        if arg == "--" { break; }
        if arg == "-1" {
            flag1 = false;
        } else if arg == "-2" {
            flag2 = false;
        } else if arg == "-3" {
            flag3 = false;
        } else if arg.starts_with('-') && arg.len() > 1 {
            eprintln!("comm: invalid option: {}", arg);
            return 1;
        } else {
            files.push(arg);
        }
    }

    if files.len() < 2 {
        eprintln!("comm: missing operand");
        return 1;
    }
    if files.len() > 2 {
        eprintln!("comm: extra operand: {}", files[2]);
        return 1;
    }

    let file1: Box<dyn BufRead> = if files[0] == "-" {
        Box::new(BufReader::new(io::stdin()))
    } else {
        match File::open(files[0]) {
            Ok(f) => Box::new(BufReader::new(f)),
            Err(e) => {
                eprintln!("comm: {}: {}", files[0], e);
                return 1;
            }
        }
    };

    let file2: Box<dyn BufRead> = if files[1] == "-" {
        Box::new(BufReader::new(io::stdin()))
    } else {
        match File::open(files[1]) {
            Ok(f) => Box::new(BufReader::new(f)),
            Err(e) => {
                eprintln!("comm: {}: {}", files[1], e);
                return 1;
            }
        }
    };

    let mut lines1 = file1.lines();
    let mut lines2 = file2.lines();

    let mut line1: Option<String> = None;
    let mut line2: Option<String> = None;

    loop {
        if line1.is_none() {
            match lines1.next() {
                Some(Ok(l)) => line1 = Some(l),
                Some(Err(e)) => {
                    eprintln!("comm: read error: {}", e);
                    return 1;
                }
                None => {}
            }
        }
        if line2.is_none() {
            match lines2.next() {
                Some(Ok(l)) => line2 = Some(l),
                Some(Err(e)) => {
                    eprintln!("comm: read error: {}", e);
                    return 1;
                }
                None => {}
            }
        }

        match (line1.take(), line2.take()) {
            (None, None) => break,
            (Some(l1), None) => {
                if flag1 {
                    pwriteln!(w, "{}", l1);
                }
                // keep consuming l1
                // Continue reading remaining lines1
                for line in &mut lines1 {
                    match line {
                        Ok(l) => {
                            if flag1 {
                                pwriteln!(w, "{}", l);
                            }
                        }
                        Err(_) => break,
                    }
                }
                break;
            }
            (None, Some(l2)) => {
                if flag2 {
                    pwriteln!(w, "\t{}", l2);
                }
                for line in &mut lines2 {
                    match line {
                        Ok(l) => {
                            if flag2 {
                                pwriteln!(w, "\t{}", l);
                            }
                        }
                        Err(_) => break,
                    }
                }
                break;
            }
            (Some(l1), Some(l2)) => {
                match l1.cmp(&l2) {
                    std::cmp::Ordering::Less => {
                        if flag1 {
                            pwriteln!(w, "{}", l1);
                        }
                        line2 = Some(l2); // keep l2 for next comparison
                        line1 = None;
                    }
                    std::cmp::Ordering::Greater => {
                        if flag2 {
                            pwriteln!(w, "\t{}", l2);
                        }
                        line1 = Some(l1); // keep l1 for next comparison
                        line2 = None;
                    }
                    std::cmp::Ordering::Equal => {
                        if flag3 {
                            pwriteln!(w, "\t\t{}", l1);
                        }
                        line1 = None;
                        line2 = None;
                    }
                }
            }
        }
    }

    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_comm_no_args() {
        let mut buf = Vec::new();
        let code = run(&mut buf, &[]);
        assert_eq!(code, 1);
    }

    #[test]
    fn test_comm_one_arg() {
        let mut buf = Vec::new();
        let code = run(&mut buf, &["/dev/null".into()]);
        assert_eq!(code, 1);
    }

    #[test]
    fn test_comm_invalid_option() {
        let mut buf = Vec::new();
        let code = run(&mut buf, &["-x".into(), "/dev/null".into(), "/dev/null".into()]);
        assert_eq!(code, 1);
    }
}
